### PR TITLE
Refactor splitOnSuffixSeq

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 1]
```diff
@@ -2075,258 +2071,341 @@ splitSuffixOn
     -> Fold m a b
     -> Stream m a
     -> Stream m b
-splitSuffixOn withSep patArr (Fold fstep initial done)
-                (Stream step state) =
-    Stream stepOuter GO_START1
+splitSuffixOn withSep patArr (Fold fstep initial done) (Stream step state) =
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2074]: Change this to `splitOnSuffixSeq`

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 2]
```diff
@@ -2075,258 +2071,341 @@ splitSuffixOn
...
-        {-# INLINE go1 #-}
-        go1 !_ !cksum !rh st !acc = do
+                    else
+                        if withSep
+                        then do
+                            r <- done fs
+                            skip $ SplitOnSeqYield r SplitOnSeqDone
+                        else skip $ SplitOnSeqKRDone idx fs rb (RB.startOf rb)
+
+    -- XXX The recursive "go" is more efficient than the state based recursion
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2355]: We can remove this comment.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 3]
```diff
@@ -2246,28 +2225,15 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
...
-                        if idx == maxIndex
-                        then do
-                            if wrd1 .&. wordMask == wordPat
-                            then do
-                                r <- done fs
-                                let next = SplitOnSuffixSeqWordInit s
-                                skip $ SplitOnSuffixSeqYield r next
-                            else skip $ SplitOnSuffixSeqWordLoop wrd1 s fs1
-                        else go SPEC (idx + 1) wrd1 s fs1
+                    fs1 <- if withSep then fstep fs x else return fs
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2228]: Can you change this back to normal, I've written it like such with partial folds in mind.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 4]
```diff
@@ -2109,6 +2109,19 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
 
     addToWord wd a = (wd `shiftL` elemBits) .|. fromIntegral (fromEnum a)
 
+    -- For single element pattern case
+    {-# INLINE processYieldSingle #-}
+    processYieldSingle pat x s fs =
+        if pat == x
+        then do
+            fs1 <- if withSep then fstep fs x else return fs
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2117]: Please move the `if` statement outside.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 5]
```diff
@@ -2284,33 +2250,25 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
                     let wrd1 = addToWord wrd x
                         old = (wordMask .&. wrd)
                                 `shiftR` (elemBits * (patLen - 1))
-                    if withSep
+                    fs1 <-
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2253]: Same as above.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 6]
```diff
@@ -2322,16 +2280,13 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
             Yield x s -> do
                 rh1 <- liftIO $ RB.unsafeInsert rb rh0 x
                 fs <- initial
-                if withSep
-                then do
-                    fs1 <- fstep fs x
-                    skip $ SplitOnSuffixSeqKRInit1 fs1 s rb rh1
-                else skip $ SplitOnSuffixSeqKRInit1 fs s rb rh1
+                fs1 <- if withSep then fstep fs x else return fs
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2283]: Same as above.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 7]
```diff
@@ -2340,45 +2295,26 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
...
-                        if idx == maxIndex
-                        then do
-                            let fold = RB.unsafeFoldRing (RB.ringBound rb)
-                            let !ringHash = fold addCksum 0 rb
-                            if ringHash == patHash
-                            then skip $ SplitOnSuffixSeqKRCheck fs s rb rh1
-                            else skip $ SplitOnSuffixSeqKRLoop
-                                            fs s rb rh1 ringHash
-                        else go SPEC (idx + 1) rh1 s fs
+                    fs1 <- if withSep then fstep fs x else return fs
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2298]: Same as above.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 8]
```diff
@@ -2391,27 +2327,19 @@ splitOnSuffixSeq withSep patArr (Fold fstep initial done) (Stream step state) =
...
-                        fs1 <- fstep fs x
-                        if (cksum1 == patHash)
-                        then skip $ SplitOnSuffixSeqKRCheck fs1 s rb rh1
-                        else go SPEC fs1 s rh1 cksum1
-                    else do
-                        fs1 <- fstep fs old
-                        if (cksum1 == patHash)
-                        then skip $ SplitOnSuffixSeqKRCheck fs1 s rb rh1
-                        else go SPEC fs1 s rh1 cksum1
+                    fs1 <- if withSep then fstep fs x else fstep fs old
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:2330]: Same as above.

File: src/Streamly/Internal/Data/Stream/StreamD.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:2228]: I do not see how partial folds will change things fundamentally. You could just change `return fs` to `return $ Partial 0 fs` and the rest all should be handled in the same way. Am I missing something?

Comment (adithyaov) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:2228]: Yeah, I want to write code in a very idiomatic and straightforward way. Once it has the same performance, I'll abstract things out bit by bit.
The last time I tried to abstract this out, the benchmarks were not the prettiest. I made a few other changes then, so technically cannot blame it on this. I'll reduce duplication as much as possible in the end.

Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:2228]: I will suggest that you can change it to whatever form when you make those changes. At this time we should write whatever is best now.

**Comment by adithyaov** [comment]
I'm merging this.

**Comment by adithyaov** [comment]
@harendra-kumar Or you can go ahead and merge it and make any final changes that you want to.

**Comment by adithyaov** [comment]
Ok, I'll take the liberty of merging this then.

**Pull request merged**